File: src/web_server/state.rs
use core::f32;
use std::str::FromStr;

use crate::data_loader::fetch_data;
use chrono::{Timelike, Utc};
use chrono_tz::Europe::Prague;
use dashmap::DashMap;
use serde::Serialize;
use tokio::join;

use super::conditions::EvaluateContext;

#[derive(Serialize, Clone)]
pub struct DayPrices {
    pub prices: Vec<f32>,
}

pub trait PriceStats {
    fn cheapest_hour(&self) -> (usize, &f32);
    fn expensive_hour(&self) -> (usize, &f32);
}

impl<'a> PriceStats for &'a [f32] {
    fn cheapest_hour(&self) -> (usize, &f32) {
        self.iter()
            .enumerate()
            .min_by(|(_, a), (_, b)| a.partial_cmp(b).unwrap())
            .unwrap()
    }

    fn expensive_hour(&self) -> (usize, &f32) {
        self.iter()
            .enumerate()
            .max_by(|(_, a), (_, b)| a.partial_cmp(b).unwrap())
            .unwrap()
    }
}

impl DayPrices {
    pub fn total_prices(&self, dist: &Distribution) -> Vec<f32> {
        let mut prices = self.prices.clone();
        for (i, price) in prices.iter_mut().enumerate() {
            if dist.high_hours.contains(&(i as u8 / 4)) {
                *price += dist.high_price;
            } else {
                *price += dist.low_price;
            }
        }

        prices
    }
}

pub struct Distribution {
    pub high_hours: Vec<u8>,
    pub high_price: f32,
    pub low_price: f32,
}

impl Distribution {
    pub fn by_hours(&self) -> [&str; 24] {
        let mut distribution = ["N"; 24];
        for hour in self.high_hours.iter() {
            distribution[*hour as usize] = "V";
        }
        distribution
    }
}

pub struct AppState {
    pub days: DashMap<chrono::NaiveDate, DayPrices>,
    pub distribution: Distribution,
}

const NEXT_DAY_PRICES_HOUR: u32 = 13;

impl AppState {
    pub fn new() -> Self {
        Self {
            days: DashMap::new(),
            distribution: Distribution {
                high_hours: vec![10, 12, 14, 17],
                high_price: 648.0 / 25.29,
                low_price: 438.0 / 25.29,
            },
        }
    }
    pub async fn get_prices(&self, date: &chrono::NaiveDate) -> Option<DayPrices> {
        if !self.days.contains_key(date) {
            match fetch_data(*date).await {
                Ok(prices) => {
                    self.days.insert(*date, DayPrices { prices: prices.clone() });

                    return Some(DayPrices { prices });
                }
                Err(_) => {
                    return None;
                }
            }
        }

        self.days.get(date).map(|i| i.value().clone())
    }

    pub async fn expression_context(&self) -> Option<EvaluateContext> {
        let now = Utc::now().with_timezone(&Prague);
        let hour = now.time().hour();

        let today = now.date_naive();
        let yesterday = today - chrono::Duration::days(1);
        let tomorrow = today + chrono::Duration::days(1);

        let join_yesterday = self.get_prices(&yesterday);
        let join_today = self.get_prices(&today);
        let join_tomorrow = self.get_prices(&tomorrow);

        let prices: (Option<DayPrices>, Option<DayPrices>, Option<DayPrices>) =
            if hour >= NEXT_DAY_PRICES_HOUR {
                join!(join_yesterday, join_today, join_tomorrow)
            } else {
                let (yesterday, today) = join!(join_yesterday, join_today);

                (yesterday, today, None)
            };

        match prices {
            (yesterday, Some(today), tomorrow) => {
                let mut prices: Vec<f32> = Vec::new();
                let mut offset = 0;

                if let Some(yesterday) = yesterday {
                    prices.extend_from_slice(&yesterday.prices);
                    offset = 24;
                }

                prices.extend_from_slice(&today.prices);

                if let Some(tomorrow) = tomorrow {
                    prices.extend_from_slice(&tomorrow.prices);
                }

                Some(EvaluateContext::new(
                    now.naive_local(),
                    prices,
                    (hour + offset) as usize,
                ))
            }
            _ => None,
        }
    }
}

#[derive(Clone, Copy, Debug)]
pub enum Currency {
    Eur,
    Czk,
}

impl Currency {
    pub const RATE: f32 = 24.30;

    pub fn convert(self, price: f32) -> f32 {
        match self {
            Currency::Eur => price,
            Currency::Czk => price * Self::RATE / 1000.0,
        }
    }

    pub fn label(self) -> &'static str {
        match self {
            Currency::Eur => "Price EUR/MWh",
            Currency::Czk => "Price CZK/kWh",
        }
    }

    pub fn short_label(self) -> &'static str {
        match self {
            Currency::Eur => "EUR/MWh",
            Currency::Czk => "CZK/kWh",
        }
    }
}

impl FromStr for Currency {
    type Err = String;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        match s.to_lowercase().as_str() {
            "eur" => Ok(Currency::Eur),
            "czk" => Ok(Currency::Czk),
            _ => Err(format!("unknown currency: {s}")),
        }
    }
}

impl std::fmt::Display for Currency {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Currency::Eur => write!(f, "eur"),
            Currency::Czk => write!(f, "czk"),
        }
    }
}

File: webserver/state.go
package webserver

import (
	"fmt"
	"log"
	"math"
	"strings"
	"sync"
	"time"

	"github.com/MichalKalita/ote/dataloader"
	"github.com/MichalKalita/ote/storage"
)

type DayPrices struct {
	Prices []float32
}

// CheapestHour returns the index and the value of the lowest price.
func CheapestHour(prices []float32) (int, float32) {
	idx := 0
	min := float32(math.Inf(1))
	for i, p := range prices {
		if p < min {
			min = p
			idx = i
		}
	}
	return idx, min
}

// ExpensiveHour returns the index and the value of the highest price.
func ExpensiveHour(prices []float32) (int, float32) {
	idx := 0
	max := float32(math.Inf(-1))
	for i, p := range prices {
		if p > max {
			max = p
			idx = i
		}
	}
	return idx, max
}

// TotalPrices returns prices including distribution surcharges.
func (d *DayPrices) TotalPrices(dist *Distribution) []float32 {
	out := make([]float32, len(d.Prices))
	for i, price := range d.Prices {
		hour := byte(i / 4)
		if containsByte(dist.HighHours, hour) {
			out[i] = price + dist.HighPrice
		} else {
			out[i] = price + dist.LowPrice
		}
	}
	return out
}

func containsByte(s []byte, v byte) bool {
	for _, x := range s {
		if x == v {
			return true
		}
	}
	return false
}

type Distribution struct {
	HighHours []byte
	HighPrice float32
	LowPrice  float32
}

// ByHours returns an array of 24 labels ("V" for high, "N" for low).
func (d *Distribution) ByHours() [24]string {
	var out [24]string
	for i := range out {
		out[i] = "N"
	}
	for _, h := range d.HighHours {
		out[h] = "V"
	}
	return out
}

type AppState struct {
	db           *storage.DB
	Distribution Distribution
}

const NextDayPricesHour = 14

func NewAppState(db *storage.DB) *AppState {
	return &AppState{
		db: db,
		Distribution: Distribution{
			HighHours: []byte{10, 12, 14, 17},
			HighPrice: 648.0 / 25.29,
			LowPrice:  438.0 / 25.29,
		},
	}
}

// GetPrices returns prices for the date. Reads from the DB; if absent, fetches
// from OTE and persists. Returns (nil, false) on fetch error.
func (s *AppState) GetPrices(date time.Time) (*DayPrices, bool) {
	pragueDate := s.db.PragueDate(date)

	has, err := s.db.HasDay(pragueDate)
	if err != nil {
		log.Printf("HasDay(%s) error: %v", pragueDate, err)
		return nil, false
	}

	if !has {
		quarters, err := dataloader.FetchData(date)
		if err != nil {
			return nil, false
		}
		if err := s.db.SaveQuarters(quarters); err != nil {
			log.Printf("SaveQuarters(%s) error: %v", pragueDate, err)
			return nil, false
		}
		return &DayPrices{Prices: quartersToPrices(quarters)}, true
	}

	quarters, err := s.db.GetDay(pragueDate)
	if err != nil {
		log.Printf("GetDay(%s) error: %v", pragueDate, err)
		return nil, false
	}
	return &DayPrices{Prices: quartersToPrices(quarters)}, true
}

func quartersToPrices(quarters []storage.Quarter) []float32 {
	out := make([]float32, len(quarters))
	for i, q := range quarters {
		out[i] = q.Price
	}
	return out
}

// MonthAverages returns daily averages keyed by day-of-month (1..31) for the
// given Prague-local month. Days strictly after maxDate are skipped. Missing
// days are fetched and persisted on first access; subsequent calls hit only
// the DB.
func (s *AppState) MonthAverages(year int, month time.Month, loc *time.Location, _ bool, maxDate time.Time) map[int]float32 {
	first := time.Date(year, month, 1, 0, 0, 0, 0, loc)
	daysInMonth := first.AddDate(0, 1, -1).Day()

	var wg sync.WaitGroup
	for day := 1; day <= daysInMonth; day++ {
		d := time.Date(year, month, day, 0, 0, 0, 0, loc)
		if d.After(maxDate) {
			continue
		}
		wg.Add(1)
		go func(d time.Time) {
			defer wg.Done()
			s.GetPrices(d)
		}(d)
	}
	wg.Wait()

	from := time.Date(year, month, 1, 0, 0, 0, 0, loc).Format("2006-01-02")
	to := time.Date(year, month, daysInMonth, 0, 0, 0, 0, loc).Format("2006-01-02")
	avgs, err := s.db.MonthAverages(from, to)
	if err != nil {
		log.Printf("MonthAverages SELECT error: %v", err)
		return map[int]float32{}
	}

	out := make(map[int]float32, len(avgs))
	for date, avg := range avgs {
		t, err := time.ParseInLocation("2006-01-02", date, loc)
		if err != nil {
			continue
		}
		out[t.Day()] = avg
	}
	return out
}

// ExpressionContext builds an EvaluateContext from yesterday/today (+tomorrow if late enough).
func (s *AppState) ExpressionContext() *EvaluateContext {
	loc, err := time.LoadLocation("Europe/Prague")
	if err != nil {
		loc = time.UTC
	}
	now := time.Now().In(loc)
	hour := now.Hour()
	today := time.Date(now.Year(), now.Month(), now.Day(), 0, 0, 0, 0, loc)
	yesterday := today.AddDate(0, 0, -1)
	tomorrow := today.AddDate(0, 0, 1)

	type fetchResult struct {
		prices *DayPrices
		ok     bool
	}
	var wg sync.WaitGroup
	var ysd, td, tmw fetchResult

	wg.Add(2)
	go func() {
		defer wg.Done()
		ysd.prices, ysd.ok = s.GetPrices(yesterday)
	}()
	go func() {
		defer wg.Done()
		td.prices, td.ok = s.GetPrices(today)
	}()
	if hour >= NextDayPricesHour {
		wg.Add(1)
		go func() {
			defer wg.Done()
			tmw.prices, tmw.ok = s.GetPrices(tomorrow)
		}()
	}
	wg.Wait()

	if !td.ok {
		return nil
	}

	var prices []float32
	offset := 0
	if ysd.ok {
		prices = append(prices, ysd.prices.Prices...)
		offset = 24
	}
	prices = append(prices, td.prices.Prices...)
	if tmw.ok {
		prices = append(prices, tmw.prices.Prices...)
	}

	nowLocal := time.Date(now.Year(), now.Month(), now.Day(), now.Hour(), now.Minute(), now.Second(), now.Nanosecond(), time.UTC)
	return NewEvaluateContext(nowLocal, prices, hour+offset)
}

type Currency int

const (
	CurrencyEur Currency = iota
	CurrencyCzk
)

const CurrencyRate float32 = 24.30

func (c Currency) Convert(price float32) float32 {
	switch c {
	case CurrencyEur:
		return price
	case CurrencyCzk:
		return price * CurrencyRate / 1000.0
	}
	return price
}

func (c Currency) ShortLabel() string {
	switch c {
	case CurrencyEur:
		return "EUR/MWh"
	case CurrencyCzk:
		return "CZK/kWh"
	}
	return ""
}

func (c Currency) String() string {
	switch c {
	case CurrencyEur:
		return "eur"
	case CurrencyCzk:
		return "czk"
	}
	return ""
}

func ParseCurrency(s string) (Currency, error) {
	switch strings.ToLower(s) {
	case "eur":
		return CurrencyEur, nil
	case "czk":
		return CurrencyCzk, nil
	}
	return CurrencyEur, fmt.Errorf("unknown currency: %s", s)
}
